File: backend/src/projects/controllers/project.controller.ts
import { Controller, Get, Post, Body, Param, HttpStatus, HttpCode } from '@nestjs/common';
import { ApiTags, ApiOperation, ApiResponse, ApiParam, ApiBody } from '@nestjs/swagger';
import { IProjectService } from '../interfaces/project.service.interface';
import { CreateProjectDto, ProjectDto } from '../dto/project.dto';

/**
 * Контроллер для управления проектами
 */
@ApiTags('projects')
@Controller('projects')
export class ProjectController {
  constructor(private readonly projectService: IProjectService) {}

  /**
   * Создает новый проект
   * @param createProjectDto Данные для создания проекта
   * @returns Созданный проект
   */
  @Post()
  @HttpCode(HttpStatus.CREATED)
  @ApiOperation({ 
    summary: 'Создание проекта',
    description: 'Создает новый проект в системе'
  })
  @ApiBody({ type: CreateProjectDto })
  @ApiResponse({ 
    status: 201, 
    description: 'Проект успешно создан',
    type: ProjectDto
  })
  @ApiResponse({ 
    status: 400, 
    description: 'Некорректные данные запроса'
  })
  async createProject(@Body() createProjectDto: CreateProjectDto): Promise<ProjectDto> {
    return this.projectService.createProject(createProjectDto);
  }

  /**
   * Получает проект по идентификатору
   * @param id Идентификатор проекта
   * @returns Проект
   */
  @Get(':id')
  @ApiOperation({ 
    summary: 'Получение проекта по ID',
    description: 'Возвращает проект по его уникальному идентификатору'
  })
  @ApiParam({ 
    name: 'id', 
    description: 'Уникальный идентификатор проекта',
    example: '123e4567-e89b-12d3-a456-426614174000'
  })
  @ApiResponse({ 
    status: 200, 
    description: 'Проект найден',
    type: ProjectDto
  })
  @ApiResponse({ 
    status: 404, 
    description: 'Проект не найден'
  })
  async getProject(@Param('id') id: string): Promise<ProjectDto | null> {
    return this.projectService.getProjectById(id);
  }

  /**
   * Получает все проекты
   * @returns Список всех проектов
   */
  @Get()
  @ApiOperation({ 
    summary: 'Получение всех проектов',
    description: 'Возвращает список всех проектов в системе'
  })
  @ApiResponse({ 
    status: 200, 
    description: 'Список проектов получен успешно',
    type: [ProjectDto]
  })
  async getAllProjects(): Promise<ProjectDto[]> {
    return this.projectService.getAllProjects();
  }
}

File: backend/src/projects/controllers/ProjectController.cs
using System;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Mvc;
using Testoria.Projects.DTO;
using Testoria.Projects.Interfaces;

namespace Testoria.Projects.Controllers
{
    /// <summary>
    /// Контроллер для управления проектами
    /// </summary>
    [ApiController]
    [Route("api/[controller]")]
    public sealed class ProjectController : ControllerBase
    {
        private readonly IProjectService _projectService;

        /// <summary>
        /// Инициализирует новый экземпляр класса ProjectController
        /// </summary>
        /// <param name="projectService">Сервис для работы с проектами</param>
        public ProjectController(IProjectService projectService)
        {
            _projectService = projectService;
        }

        /// <summary>
        /// Создает новый проект
        /// </summary>
        /// <param name="createProjectDto">Данные для создания проекта</param>
        /// <returns>Созданный проект</returns>
        /// <response code="201">Проект успешно создан</response>
        /// <response code="400">Некорректные данные запроса</response>
        [HttpPost]
        [ProducesResponseType(typeof(ProjectDto), 201)]
        [ProducesResponseType(400)]
        public async Task<ActionResult<ProjectDto>> CreateProject([FromBody] CreateProjectDto createProjectDto)
        {
            if (!ModelState.IsValid)
            {
                return BadRequest(ModelState);
            }

            var project = await _projectService.CreateProjectAsync(createProjectDto);
            return CreatedAtAction(nameof(GetProject), new { id = project.Id }, project);
        }

        /// <summary>
        /// Получает проект по идентификатору
        /// </summary>
        /// <param name="id">Идентификатор проекта</param>
        /// <returns>Проект</returns>
        /// <response code="200">Проект найден</response>
        /// <response code="404">Проект не найден</response>
        [HttpGet("{id}")]
        [ProducesResponseType(typeof(ProjectDto), 200)]
        [ProducesResponseType(404)]
        public async Task<ActionResult<ProjectDto>> GetProject(Guid id)
        {
            var project = await _projectService.GetProjectByIdAsync(id);
            if (project == null)
            {
                return NotFound();
            }

            return Ok(project);
        }

        /// <summary>
        /// Получает все проекты
        /// </summary>
        /// <returns>Список всех проектов</returns>
        /// <response code="200">Список проектов получен</response>
        [HttpGet]
        [ProducesResponseType(typeof(ProjectDto[]), 200)]
        public async Task<ActionResult<ProjectDto[]>> GetAllProjects()
        {
            var projects = await _projectService.GetAllProjectsAsync();
            return Ok(projects);
        }
    }
}
